src/coreclr/jit/lower.cpp

```diff
@@ -8003,6 +8016,64 @@ void Lowering::LowerStoreIndirCoalescing(GenTreeStoreInd* ind)
             return;
         }
 
+        // Now the hardest part: decide whether it's safe to use an unaligned write.
+        //
+        // IND<byte> is always fine (and all IND<X> created here from such)
+        // IND<simd> is not required to be atomic per our Memory Model
+        const bool allowsNonAtomic =
+            ((ind->gtFlags & GTF_IND_ALLOW_NON_ATOMIC) != 0) && ((prevInd->gtFlags & GTF_IND_ALLOW_NON_ATOMIC) != 0);
+
+        if (!allowsNonAtomic && (genTypeSize(ind) > 1) && !varTypeIsSIMD(ind))
+        {
+            // TODO-CQ: if we see that the target is a local memory (non address exposed)
+            // we can use any type (including SIMD) for a new load.
+
+            // Ignore indices for now, they can invalidate our alignment assumptions.
+            // Although, we can take scale into account.
+            if (currData.index != nullptr)
+            {
+                return;
+            }
+
+            // Base address being TYP_REF gives us a hint that data is pointer-aligned.
+            if (!currData.baseAddr->TypeIs(TYP_REF))
+            {
+                return;
+            }
+
+            const int ptrAlignOffset = min(prevData.offset, currData.offset) % genTypeSize(ind);
+
+#ifdef TARGET_XARCH
+            // Check whether data can be guaranteed to be in a cache line.
+            // Only AMD/Intel can guarantee that unaligned writes are atomic if they don't cross the
+            // cache line boundary.
+            if ((ptrAlignOffset + genTypeSize(ind) * 2) > TARGET_POINTER_SIZE)
+            {
+                // NOTE: We can use SIMD if we can assume 16-bytes alignment (on Intel/AMD cpus with AVX+),
+                // but it's very unlikely we can detect that from surrounding code since GC only offers 8-bytes one.
+                return;
+            }
```

Comment:
I'm not quite sure I understand this check, I would expect it to be
```cpp
bool isCombinedIndirAtomic = (min(prevData.offset, curData.offset) % (genTypeSize(ind) * 2)) == 0;
```

and then additionally allowing the `ind->TypeIs(TYP_LONG, TYP_REF)` cases for arm64 as below.

Reply:
@jakobbotsch ah.. hm.. can you check my recent commit if I addressed it?

Also, I realized I no longer need to set GTF_UNALIGNED on the read since we maintain the alignment.

Reply:
Looks better now, but thinking about it again I think we also need a `genTypeSize(ind) < TARGET_POINTER_SIZE` to take the base alignment into account (if I understand currently the code now might combine 2x `TYP_I_IMPL` into `TYP_SIMD16` on xarch)

Reply:
Sorry, I guess the check actually needs to be `!ind->TypeIs(TYP_I_IMPL)`

Reply:
Ah, indeed, it could also enable 2xINT to LONG on 32bit.

I think `genTypeSize(ind) < TARGET_POINTER_SIZE` reads better to me